### ISSUE TITLE
Make vapoursynth a standard Python package

### ISSUE THREAD

**Issue by chengyuhui** [vapoursynth/vapoursynth#283]
The current installation method involves running an installer or copying files into python directory, which makes it difficult to deploy applications and upgrading python automatically.

Is it possible to use setuptools to do the packaging?

**Comment by stuxcrystal** [comment]
The main problem is here: `setup.py` L12-17
```python
library_dirs = [curdir, "build"]
if is_win:
    if is_64:
        library_dirs.append(join("msvc_project", "x64", "Release"))
    else:
        library_dirs.append(join("msvc_project", "Release"))
```

It hardcodes the places where the vapoursynth.lib is being searched.

**Comment by Kamekameha** [comment]
Yeah, agree. IMO the `setup.py` script shouldn't have any hardcoded paths and let the user building the python module specify the include and library directories manually (with the `-I`/`-L`  options accordingly, or in `CFLAGS`/`LDFLAGS` if the system supports it). Though that basically means one extra thing to keep in mind for those distributing this, but that's essentially how all CPython extensions are distributed on PyPI.

**Comment by stuxcrystal** [comment]
I chatted with Myrsloik yesterday. I am going to add a bit of code adding the discovery of the vapoursynth-sdk on windows. On other systems, I will just assume it will be on the path.

**Comment by chengyuhui** [comment]
Actually, I'm thinking of moving everything (vapoursynth itself, plugins, scripts, etc.) onto PyPI.

AFAIK, not many vapoursynth users are real Python developers, but we can still benefit from the Python ecosystem. 

e.g.: Once this is done, it'll be easier to implement #52 

Another problem: is there any real difference between plugins and core plugins? Maybe we can distribute them separately.

**Comment by myrsloik** [comment]
Core plugins are always guaranteed to be present and should never be touched by the user. They're just plugins to keep the structure clean.

**Comment by splinter98** [comment]
I have thought about this (and similar related) issues a fair amount over the years in my spare time. 

The main issue I have with moving everything to PyPi is the core of Vapoursynth isn't Python. Only the (supported) front end.

Bar a few help modules all plugins are C++ and wouldn't package well inside python. Moving Vapoursynth to PyPi would require restructuring the whole VS ecosystem, which while I'm not adverse to, in this case I just can't see it being beneficial. For example when using VS/AVFS or VSEditor or other third party utilities.

I can  however potentially see some merit in the VScript portion being on PyPi and the installer calling pip to manage that part, and for python modules they definitely could today be distributed over PyPi. This would allow users to be able to install vapoursynth in their virtual environments or non standard installations much easier.

The way I see it Vapoursynth embeds python rather than Vapoursynth being a Python based application. Remember Vapoursynth is written in a way you could (and it has been done) change the front end language to anything you want ( [vsjs](https://forum.doom9.org/showthread.php?t=173869&highlight=javascript) for example enables you to use all the C++ plugins but write up your wiring code in Javascript instead of Python).

A better ecosystem to follow would be to do something similar to what is being done with the Atom editor project. I would have a either an copy of python embedded or utilizing a virtual environment and keep all VS specific stuff outside of system Python. This would also help reduce some of the Linux distro complexities we have had to deal with in the past.

This doesn't remove the requirement of running an installer but it will mean it won't conflict with python being upgraded as it is using it's own internal version.

**Comment by chengyuhui** [comment]
@splinter98 Thanks for your reply.

If I'm understanding correctly, what we are really using is a Python binding of the VapourSynth library.
Then we would rather have our own package management system, etc.

We can just ship the Python binding on PyPI, then make it possible for the binding to find a VapourSynth installation automatically.

Though we still need an installer, we won't need to touch Python installation, making it difficult to upgrade it.

I have tried to put portable version into an separate directory instead of extracting into an Python installation, it worked, but the code is quite ugly. ( Unlike in Node.js, we can't load a module only with its path easily with `require('./vapoursynth');`)

**Comment by splinter98** [comment]
> If I'm understanding correctly, what we are really using is a Python binding of the VapourSynth library. Then we would rather have our own package management system, etc.

Yes, effectively that is the case.

> We can just ship the Python binding on PyPI, then make it possible for the binding to find a VapourSynth installation automatically.

It's still a little tricky, you still need your application to invoke the correct vsscript library and in turn the correct python library, and it's all done currently relative to the original calling executable's library location. But it is doable.

I have spoken about how you can set up a virtualenv to do this previously (setup should be similar for portable mode) here: https://github.com/vapoursynth/vapoursynth/issues/156#issuecomment-86158364

**Comment by Kamekameha** [comment]
One way could be to move VSScript functionality entirely to the python module. For example, numpy exposes a C API (which is essentially equivalent to what `vapoursynth_api.h` is for) and ship the library to link to right away in their package. The python bindings of vapoursynth could do something similar to this, but then this will kill the point (as I see it as) of VSScript being a convenient wrapper around the already exposed C API of the python bindings, that calls CPython specific code to initialize it and etc. So yeah this is a bit tricky.

As of the `setup.py` itself, @stuxcrystal you could assume the same wrt the path, the setup could be as simple as:
```python
#!/usr/bin/env python

from setuptools import Extension, setup


setup(
    name='VapourSynth',
    author='Fredrik Mellbin',
    author_email='fredrik.mellbin@gmail.com',
    version='1.0.0',
    description='A frameserver for the 21st century',
    url='http://www.vapoursynth.com/',
    download_url='https://github.com/vapoursynth/vapoursynth',
    license='LGPL 2.1 or later',
    long_description='A portable replacement for Avisynth',
    zip_safe=False,
    platforms='any',
    setup_requires=[
        'cython>=0.24'
    ],
    ext_modules=[
        Extension('vapoursynth', ['src/cython/vapoursynth.pyx'],
                  libraries=['vapoursynth'])
    ]
)
```
and it'll compile fine as long as the compiler finds the required directories, probably no need to special case for some systems on the setup file.

**Comment by stuxcrystal** [comment]
The special cases on windows were added to let the setup.py find the compiled files in the Release-directory of the MSVC Solution.

I planned on adding code to determine the vapoursynth installation path on windows system (the path is written into the Registry).

**Comment by Kamekameha** [comment]
What if users have the portable vapoursynth "installed"? Special casing would never be enough. The Release-directory stuff of the MSVC solution could be written in the `cython_build.bat` script though, now that I think about it (as of options to the `setup.py` script, like I said above).

**Comment by splinter98** [comment]
Reflecting again on this, it appears to me we have several issues in this issue:

1. - Installing Vapoursynth from source currently requires multiple steps
2. - Packaging VSScript on Windows only works when building from source

Whilst @stuxcrystal has a fix in plan for the latter, this doesn't solve the issue for the former.

What we need to do is add the installation steps into autotools for Linux and the MSVC project for windows with configuration parameters to allow advance user to select the correct version of Python & path.

**Comment by myrsloik** [comment]
@splinter98 you can currently select which python installation to install to under windows assuming it follows the PEP about python installation paths in the registry. Are there still python 3.6 distributions that don't do this properly?

And making it a python package is hard since there are so many more parts. Definitely not python only...

**Comment by splinter98** [comment]
your right, at installation time we save the location to the registry and we load python from that location. (I somehow misread that from vsscript). I don't believe there are any of the popular distributions of python on Windows that don't follow this.

There is definitely a clear common misunderstanding on the structure of Vapoursynth I have found, (and have fallen initially into myself) when coming from a python background. Vapoursynth is not a Python application at it's core, and nor should it be.

**Comment by Kamekameha** [comment]
It's totally feasible though to isolate the python bindings library from the core library, making it easier to distribute each "separately" (as in possible to install the python bindings through the usual pip or something).

As far as I can see, the main thing stopping this is vsscript (and by extension vspipe, and whatever else using it) mostly because it's compiled separately from the python extension while still relying on python. In my opinion vsscript's a bit redundant given that most of the functions there simply call its cython-exported function equivalent without doing anything else in particular, and even more redundant if the user initialized python manually beforehand, which is the only real thing (again, as far as I can see) that vsscript wraps. Users ideally should simply use the `vapoursynth.h`/`vapoursynth_api.h` headers cython exports.

**Comment by chengyuhui** [comment]
Anyway, is it possible to load VapourSynth.dll from alternative location?

**Comment by stuxcrystal** [comment]
Since VapourSynth now managed to piss me off enough to just do it, I noticed that the setup.py warns me that there is no README in this project.

I think, adding a README would be advisible for any project.

**Comment by stuxcrystal** [comment]
Implemented by #320
